mantis-server/mantis-server-worker/src/main/java/io/mantisrx/server/worker/SubscriptionStateHandlerImpl.java

```diff
@@ -0,0 +1,117 @@
+    }
+
+    @Override
+    public synchronized void start() {
+        startedAt = System.currentTimeMillis();
+        onSinkUnsubscribed(); // start off as unsubscribed
+    }
+
+    private long evalSubscriberTimeoutSecs() {
+        return Math.max(
+                minRuntimeSecs - ((System.currentTimeMillis() - startedAt) / 1000L),
+                subscriptionTimeoutSecs
+        );
+    }
+
+    @Override
+    public synchronized void onSinkUnsubscribed() {
+        if (executor == null)
+            return;
+        if (timedOutExitFutureRef.get() == null) {
+            timedOutExitFutureRef.set(
+                    executor.schedule(
+                            () -> {
+                                final AtomicReference<Throwable> error = new AtomicReference<Throwable>();
+                                while (true) {
+                                    logger.info("Calling master to kill due to subscription timeout");
+                                    final Subscription subscription = masterClientApi.killJob(jobId, "MantisWorker", "No subscriptions for " +
+                                            subscriptionTimeoutSecs + " secs")
+                                            .subscribe();
+                                    // wait for kill to happen, we won't be running if it succeeds
+                                    try {Thread.sleep(60000);} // arbitrary sleep before we retry the kill
+                                    catch (InterruptedException ie) {
+                                        logger.info("Interrupted while waiting to kill job upon timeout, cancelling");
+                                        return;
+                                    }
+                                    subscription.unsubscribe(); // unsubscribe from previous one before retrying
+                                }
+                            },
+                            evalSubscriberTimeoutSecs(), TimeUnit.SECONDS
+                    )
+            );
+            logger.info("Setup future job kill (in " + subscriptionTimeoutSecs +
+                    " secs) upon no subscribers for ephemeral job " + jobId);
+        }
+    }
+
+    @Override
+    public synchronized void onSinkSubscribed() {
+        if (executor == null) {
+            return;
+        }
+        final ScheduledFuture prevFutureKill = timedOutExitFutureRef.getAndSet(null);
+        if (prevFutureKill != null && !prevFutureKill.isCancelled() && !prevFutureKill.isDone()) {
+            logger.info("Cancelled future kill upon active subscriptions of ephemeral job " + jobId);
+            prevFutureKill.cancel(true);
+        }
+    }
+
+    @Override
+    public synchronized void shutdown() {
+        // we need to clean up the resources. onSinkSubscribed works in the same fashion.
+        // let's call that instead.
+        onSinkSubscribed();
+        // let's shutdown the executor and reset it.
+        executor.shutdown();
```

Comment:
there could be a NPE here that needs to be handled.

Reply:
Cleaned up.